activerecord/lib/active_record/attributes.rb

```diff
@@ -243,7 +243,9 @@ def _default_attributes # :nodoc:
         @default_attributes ||= begin
           # TODO: Remove the need for a connection after we release 8.1.
           attributes_hash = with_connection do |connection|
+            connection_columns = connection.schema_cache.columns_hash(table_name)
             columns_hash.transform_values do |column|
+              column = connection_columns[column.name] || column
```

Comment:
`@columns_hash` is[ set after doing some filtering on the schema cache columns on schema load](https://github.com/rails/rails/blob/0209bf4741d22c08de05ccd9fc8b14873698e14a/activerecord/lib/active_record/model_schema.rb#L592-L594), so we should iterate on the filtered set while still retrieving columns from the schema cache.